### ISSUE TITLE
Modal accessibility for WCAG 2.0

### ISSUE THREAD

**Issue by LaurentGruber** [claroline/Distribution#204]
Réfléchir à l'opportunité d'utiliser un outil tel que :

https://github.com/humaan/Modaal
http://humaan.com/modaal/

**Comment by LaurentGruber** [comment]
@om64 Qu'en penses tu ?

**Comment by stefk** [comment]
I may be wrong but I think the bootstrap accessibility plugin @ptsavdar has added to the dependencies already deals with those issues.

**Comment by om64** [comment]
It only covers dialog windows, unfortunately, but this could be a very good beginning when dealing with Claroline Connect accessibility. I don't know the bootstrap accessibility plugin.

**Comment by om64** [comment]
@LaurentGruber - J'ai tenté d'installer le plugin de ptsavdar, mais je n'y suis pas arrivé. A essayer, sans doute avant de "mettre les mains dans le cambouis".

**Comment by ptsavdar** [comment]
Hello, though I think bootstrap accessibility plugin covers more ground than modals (which is what modaal plugin does), however it wasn't me who included the plugin to the dependencies. This plugin (bootstrap accessibility) was also recommended by the Orange accessibility team who are already using it in addition to a plugin they developed called Boosted [github link](https://github.com/Orange-OpenSource/Orange-Boosted-Bootstrap) and [page](http://orange-opensource.github.io/Orange-Boosted-Bootstrap/). From what I've seen they are depending on bootstrap and bootstrap accessibility plugin and developing their theme. Which only makes me wonder what if they decide to stop supporting it? What happens then? While bootstrap moves forward to v4 (then to v5 I'm guessing...) I don't know what is the future for Boosted. Or if it would be compatible with future versions of bootstrap. And I think that's pretty much everything I know.

**Comment by stefk** [comment]
> I think bootstrap accessibility plugin covers more ground than modals

Detailed features are listed [here](https://github.com/paypal/bootstrap-accessibility-plugin#details). From what I see, simply including the plugin on the page will "fix" the markup by adding various accessibility attributes. That's interesting, though it simply does what developers and template authors should have done in the first place IMO.

> Which only makes me wonder what if they decide to stop supporting it?

The accessibility plugin is maintained by PayPal, so I think we can expect a reasonnable degree of support.


I guess the real question is: do we depend on extra tools to magically hande accessibility for us or do we learn how to detect/fix existing issues and write better markup? If we go the second route, there's also plenty of tools that can help us. I've recently found [this one](http://squizlabs.github.io/HTML_CodeSniffer/), which helps finding violations of WCAG 2.0. It's used as a bookmarklet and can work on any web page:

![capture du 2016-04-28 10 10 52](https://cloud.githubusercontent.com/assets/880921/14879679/ec992972-0d2a-11e6-979a-e78298f546fb.png)
![capture du 2016-04-28 10 11 28](https://cloud.githubusercontent.com/assets/880921/14879684/f40ffe4c-0d2a-11e6-940d-801e081a46f9.png)
![capture du 2016-04-28 10 13 26](https://cloud.githubusercontent.com/assets/880921/14879690/fc9a2fce-0d2a-11e6-8d4f-402475d827be.png)

**Comment by ptsavdar** [comment]
@stefk I agree with you that it should be a developer's (or template designer's) responsibility to include these attributes in the templates but the plugin could (at least in the beginning) offer a quick solution to the problem. 

> The accessibility plugin is maintained by PayPal, so I think we can expect a reasonnable degree of support.

My wonder about support was not for the accessibility plugin but for the Orange Boosted (in the case we decided to adopt it since it's more complete than bootstrap accessibility).

> I've recently found this one, which helps finding violations of WCAG 2.0.

The tool you suggest looks really helpful.

**Comment by om64** [comment]
@ptsavdar @stefk @LaurentGruber -  Thank you for the reactions. We've also found Chrome or Firefox extensions to check accessibility. They all seem to work well, at least partially. Never forget that there are different levels of accessibility and that websites need to be 100% accessible. I think we should use tools to check accessibility but I don't like the idea of depending on "extra tools to magically handle accessibility for us" as stefk puts it.

**Comment by om64** [comment]
@ptsavdar @stefk @LaurentGruber - 
![cvg_homepage](https://cloud.githubusercontent.com/assets/14308826/14881675/5558d282-0d35-11e6-96b2-651898ef2fab.png)

I've just tried Sniffer on CVG Home page to check level AA: interesting!

**Comment by stefk** [comment]
FYI, I've just tested the accessibility plugin and it doesn't seem to bring any improvements. In fact, the whole plugin may just be obsolete:

https://github.com/paypal/bootstrap-accessibility-plugin/issues/74#issuecomment-82491223
https://github.com/paypal/bootstrap-accessibility-plugin/issues/81

We might as well remove it from the dependencies...

PS: @om64, if you find HTML_CodeSniffer good enough, we'd probably be able to automate the checking process and make full audits of the platform using tools like https://github.com/crawlkit/runner-htmlcs.

**Comment by om64** [comment]
I've tested Code Sniffer a little further: it gives me a lot of unjustified warnings such as missing alt attributes for pictures. Furthermore, it doesn't indicate which lines of code you've got to check.

**Comment by stefk** [comment]
> it gives me a lot of unjustified warnings such as missing alt attributes for pictures

You mean the alt attribute is there and it doesn't see it?

> it doesn't indicate which lines of code you've got to check.

That can't be done AFAIK, it just analyses the final HTML, not the code that produced it.

**Comment by om64** [comment]
> You mean the alt attribute is there and it doesn't see it?
That's it!

> That can't be done AFAIK, it just analyses the final HTML, not the code that produced it.
I've got the same error in both codes...

**Comment by arnaudbey** [comment]
About accessibility, dunno to what extent this is reliable but I just stumbled on it. http://khan.github.io/tota11y/#Try-it (have to click on the glasses on bottom-left)

**Comment by stefk** [comment]
Another one, designed for automated testing: https://github.com/dequelabs/axe-core.

**Comment by stefk** [comment]
Closed in favor of https://github.com/claroline/Distribution/issues/1715.
